### PR TITLE
[ENH] Add Multi Reconstruction loss to AEFCNClusterer

### PR DESCRIPTION
**Pull request by aadya940**
Mean Squared Error between encoder and decoder layers summed up for a symmetric deep auto-encoder architecture.

### CODE REVIEW

File: aeon/clustering/deep_learning/_ae_fcn.py [hunk 1]
```diff
@@ -61,7 +62,7 @@ class AEFCNClusterer(BaseDeepClusterer):
     verbose : boolean, default = False
         Whether to output extra information.
     loss : string, default="mean_squared_error"
-        Fit parameter for the keras model.
+        Fit parameter for the keras model. "multi_rec" for multiple mse loss.
```
Comment (hadifawaz1999) [inline aeon/clustering/deep_learning/_ae_fcn.py:65]: could you detail more whats the difference between both losses in the doc ?

File: aeon/clustering/deep_learning/_ae_fcn.py [hunk 2]
```diff
@@ -303,6 +315,129 @@ def _score(self, X, y=None):
         latent_space = self.model_.layers[1].predict(X)
         return self.clusterer.score(latent_space)
 
+    def _fit_multi_rec_model(
```
Comment (hadifawaz1999) [inline aeon/clustering/deep_learning/_ae_fcn.py:318]: even though its a private function, i think its good if we add a docstirng to it, if new contributors come along they can see what the function does in the doc

File: aeon/clustering/deep_learning/_ae_fcn.py [hunk 3]
```diff
@@ -303,6 +315,129 @@ def _score(self, X, y=None):
...
+        inputs,
+        outputs,
+        batch_size,
+        epochs,
+    ):
+        import tensorflow as tf
+        from tensorflow.keras import backend as K
+
+        train_dataset = tf.data.Dataset.from_tensor_slices((inputs, outputs))
+        train_dataset = train_dataset.shuffle(buffer_size=1024).batch(batch_size)
```
Comment (hadifawaz1999) [inline aeon/clustering/deep_learning/_ae_fcn.py:330]: shouldn't the shuffle be done at each epoch iteration instead of prior to epoch looping ? because we usually shuffle at each epoch, unless this is a functionality that i dont know about and it does what am saying :) in that case never mind, if not then it should be in the loop below

File: aeon/clustering/deep_learning/_ae_fcn.py [hunk 4]
```diff
@@ -303,6 +315,129 @@ def _score(self, X, y=None):
...
+    def _fit_multi_rec_model(
+        self,
+        autoencoder,
+        inputs,
+        outputs,
+        batch_size,
+        epochs,
+    ):
+        import tensorflow as tf
+        from tensorflow.keras import backend as K
```
Comment (hadifawaz1999) [inline aeon/clustering/deep_learning/_ae_fcn.py:327]: avoid using keras backend, for instead of K.mean use tf.reduce_mean etc. as long as this is in the loss not the graph, its much more efficient to use tensorflow tensor calculation instead of keras backend

File: aeon/clustering/deep_learning/_ae_fcn.py [hunk 5]
```diff
@@ -303,6 +315,129 @@ def _score(self, X, y=None):
...
+                _decoder_intermediate_outputs = (
+                    []
+                )  # Store embeddings of each layer in the Decoder
+
+                encoder = autoencoder.layers[1]  # Returns Functional API Models.
+                decoder = autoencoder.layers[2]  # Returns Functional API Models.
+
+                # Run the models since the below given loop misses the latent space
+                # layer which doesn't contribute to the loss.
+                logits = encoder(inputs)
```
Comment (hadifawaz1999) [inline aeon/clustering/deep_learning/_ae_fcn.py:354]: why calling the models here ?

File: aeon/clustering/deep_learning/_ae_fcn.py [hunk 6]
```diff
@@ -303,6 +315,129 @@ def _score(self, X, y=None):
...
+                if not (
+                    len(_encoder_intermediate_outputs)
+                    == len(_decoder_intermediate_outputs)
+                ):
+                    raise ValueError("The Auto-Encoder must be symmetric in nature.")
+
+                for enc_output, dec_output in zip(
+                    _encoder_intermediate_outputs, _decoder_intermediate_outputs
+                ):
+                    mse += K.mean(K.square(enc_output - dec_output))
```
Comment (hadifawaz1999) [inline aeon/clustering/deep_learning/_ae_fcn.py:384]: here it should tf not K as mentioned above

File: aeon/clustering/deep_learning/_ae_fcn.py [hunk 7]
```diff
@@ -303,6 +315,129 @@ def _score(self, X, y=None):
...
+            num_batches = 0
+            for step, (x_batch_train, y_batch_train) in enumerate(train_dataset):
+                with tf.GradientTape() as tape:
+                    # Calculate the actual loss by calling the loss function
+                    loss_func = layerwise_mse_loss(
+                        autoencoder=autoencoder,
+                        inputs=x_batch_train,
+                        outputs=y_batch_train,
+                    )
+                    loss_value = loss_func(y_batch_train, autoencoder(x_batch_train))
```
Comment (hadifawaz1999) [inline aeon/clustering/deep_learning/_ae_fcn.py:406]: can u make sure here that loss_value is actually one value and not an array ? test it locally or even add an assert

File: aeon/clustering/deep_learning/_ae_fcn.py [hunk 8]
```diff
@@ -303,6 +315,129 @@ def _score(self, X, y=None):
...
+
+                grads = tape.gradient(loss_value, autoencoder.trainable_weights)
+                self.optimizer_.apply_gradients(
+                    zip(grads, autoencoder.trainable_weights)
+                )
+
+                epoch_loss += float(loss_value)
+                num_batches += 1
+
+                if step % 200 == 0:
```
Comment (hadifawaz1999) [inline aeon/clustering/deep_learning/_ae_fcn.py:416]: i prefer if the logging happens after each epoch not each step, showing the avewrage loss over all steps

File: aeon/clustering/deep_learning/_ae_fcn.py [hunk 9]
```diff
@@ -303,6 +315,129 @@ def _score(self, X, y=None):
...
+        train_dataset = tf.data.Dataset.from_tensor_slices((inputs, outputs))
+        train_dataset = train_dataset.shuffle(buffer_size=1024).batch(batch_size)
+
+        if isinstance(self.optimizer_, str):
+            self.optimizer_ = tf.keras.optimizers.get(self.optimizer_)
+
+        history = {"loss": []}
+
+        def layerwise_mse_loss(autoencoder, inputs, outputs):
+            def loss(y_true, y_pred):
```
Comment (hadifawaz1999) [inline aeon/clustering/deep_learning/_ae_fcn.py:338]: the loss is correctly coded, however you might forgot one thing, the multi rec loss also includes what the traditional  rec loss does, which is mse between input and output of decoder, so you should add this at the end of the layer wise computation

File: aeon/clustering/deep_learning/tests/test_clusterer_features.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+
+@pytest.mark.skipif(
+    not _check_soft_dependencies(["tensorflow"], severity="none"),
+    reason="Tensorflow soft dependency not found.",
+)
+def test_multi_rec_fcn():
+    """Tests whether multi-rec loss works fine or not."""
+    X = np.random.random((100, 5, 2))
+    clst = AEFCNClusterer(n_clusters=2, n_epochs=10, loss="multi_rec")
```
Comment (hadifawaz1999) [inline aeon/clustering/deep_learning/tests/test_clusterer_features.py:17]: to make sure symmetry is correctly handled in the layer wise computation, can u set the n_filters here to somehting like [2,3,4] ? because the default is [128,256,128] which 1stly is computationally expensive for a test and 2ndly a bug can easilly be missed because of the repeated 128

**Review by hadifawaz1999 — CHANGES_REQUESTED** [review]
all looks good just some minor changes, thanks !

File: aeon/clustering/deep_learning/_ae_fcn.py (hunk 5 above)
Comment (aadya940) [reply aeon/clustering/deep_learning/_ae_fcn.py:354]: It was a need since the last layer in the encoder didn't have an activation hence, there was  a shape mismatch as the tensors never passed through the last layer. Basically, it is added so that propogation of tensors between encoder and decoder doesn't break.

File: aeon/clustering/deep_learning/_ae_fcn.py (hunk 9 above)
Comment (aadya940) [reply aeon/clustering/deep_learning/_ae_fcn.py:338]: OK, so basically add normal "mse" to the loss?

Comment (hadifawaz1999) [reply aeon/clustering/deep_learning/_ae_fcn.py:338]: yes

File: aeon/clustering/deep_learning/_ae_fcn.py (hunk 3 above)
Comment (aadya940) [reply aeon/clustering/deep_learning/_ae_fcn.py:330]: In the docs its done this way: 

https://www.tensorflow.org/guide/keras/writing_a_training_loop_from_scratch

File: aeon/clustering/deep_learning/tests/test_clusterer_features.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+)
+def test_multi_rec_fcn():
+    """Tests whether multi-rec loss works fine or not."""
+    X = np.random.random((100, 5, 2))
+    clst = AEFCNClusterer(
+        n_clusters=2, n_epochs=10, n_filters=[2, 3, 4], loss="multi_rec"
+    )
+    clst.fit(X)
+    assert (
+        clst.history["loss"][0] > clst.history["loss"][1]
```
Comment (hadifawaz1999) [inline aeon/clustering/deep_learning/tests/test_clusterer_features.py:22]: this is not always true, especially first two epochs:p its so random at the beggining, you can assert that the loss is a list for example or its length is the same as number of epochs etc.

File: aeon/clustering/deep_learning/_ae_fcn.py [hunk 12]
```diff
@@ -303,6 +318,132 @@ def _score(self, X, y=None):
...
+                    __output = _model(logits, training=True)
+                    _decoder_intermediate_outputs.append(__output)
+
+                if not (
+                    len(_encoder_intermediate_outputs)
+                    == len(_decoder_intermediate_outputs)
+                ):
+                    raise ValueError("The Auto-Encoder must be symmetric in nature.")
+
+                # # Append normal mean_squared_error
```
Comment (hadifawaz1999) [inline aeon/clustering/deep_learning/_ae_fcn.py:383]: i think you forgot to append the original mse here

Comment (hadifawaz1999) [reply aeon/clustering/deep_learning/_ae_fcn.py:383]: ah no never mind i see what you did :)

**Review by hadifawaz1999 — CHANGES_REQUESTED** [review]
left some comments

File: aeon/clustering/deep_learning/tests/test_clusterer_features.py (hunk 11 above)
Comment (hadifawaz1999) [reply aeon/clustering/deep_learning/tests/test_clusterer_features.py:22]: i would set a random state however, for the data generation and the deep mode, and compare epoch 0 with epoch 5, or even 10, just to avoid random fails in the future in CI

**Review by hadifawaz1999 — APPROVED** [review]
